### DISCUSSION TITLE
Looking for a minimal working example with reactive computed mapped array properties

### DISCUSSION THREAD

**Discussion by otiv-willem-vanhulle** [slint-ui/slint#6180] [Q&A]
In the following example, I created two components: a high-level and a low-level component. A low-level component in practice may be something like a generic table widget. The high-level component in practice may be a table of some business-specific kind of items.

```slint
struct LowLevelStruct {
    background_color: brush,
    text: string
}

component LowLevelComponent {
    in property <[LowLevelStruct]> generic_items;
    for item in generic_items:
        Rectangle {
            background: item.background-color;
            Text {
                
                text: item.text;
            }
        }

}

struct HighLevelStruct {
    text: string
}

component HighLevelComponent {
    in property <[HighLevelStruct]> red_items;

    out property<[LowLevelStruct]> generic_red_computed_items; //: red_items.map(|item| {background_color: red, text: item.text})
    LowLevelComponent {
        generic_items: root.generic_red_computed_items;
    }
}

export component Demo {
    width: 400px;
    height: 200px;
    VerticalLayout {
        LowLevelComponent {
            generic_items: [
                {
                    background_color: Colors.red,
                    text: "Item manually set to red"
                }
            ];
        }

        Text {
            text: "Below I would like to render using a 'computed' property\nin the higher-level component.";
        }

        HighLevelComponent {
            red_items: [{ text: "This should be rendered red" }];
        }
    }    
}
```

[link to playground](https://slintpad.com/?snippet=struct+LowLevelStruct+%7B%0A++++background_color%3A+brush%2C%0A++++text%3A+string%0A%7D%0A%0Acomponent+LowLevelComponent+%7B%0A++++in+property+%3C%5BLowLevelStruct%5D%3E+generic_items%3B%0A++++for+item+in+generic_items%3A%0A++++++++Rectangle+%7B%0A++++++++++++background%3A+item.background-color%3B%0A++++++++++++Text+%7B%0A++++++++++++++++%0A++++++++++++++++text%3A+item.text%3B%0A++++++++++++%7D%0A++++++++%7D%0A%0A%7D%0A%0Astruct+HighLevelStruct+%7B%0A++++text%3A+string%0A%7D%0A%0Acomponent+HighLevelComponent+%7B%0A++++in+property+%3C%5BHighLevelStruct%5D%3E+red_items%3B%0A%0A++++out+property%3C%5BLowLevelStruct%5D%3E+generic_red_computed_items%3B+%2F%2F%3A+red_items.map%28%7Citem%7C+%7Bbackground_color%3A+red%2C+text%3A+item.text%7D%29%0A++++LowLevelComponent+%7B%0A++++++++generic_items%3A+root.generic_red_computed_items%3B%0A++++%7D%0A%7D%0A%0Aexport+component+Demo+%7B%0A++++width%3A+400px%3B%0A++++height%3A+200px%3B%0A++++VerticalLayout+%7B%0A++++++++LowLevelComponent+%7B%0A++++++++++++generic_items%3A+%5B%0A++++++++++++++++%7B%0A++++++++++++++++++++background_color%3A+Colors.red%2C%0A++++++++++++++++++++text%3A+%22Item+manually+set+to+red%22%0A++++++++++++++++%7D%0A++++++++++++%5D%3B%0A++++++++%7D%0A%0A++++++++Text+%7B%0A++++++++++++text%3A+%22Below+I+would+like+to+render+using+a+%27computed%27+property%5Cnin+the+higher-level+component.%22%3B%0A++++++++%7D%0A%0A++++++++HighLevelComponent+%7B%0A++++++++++++red_items%3A+%5B%7B+text%3A+%22This+should+be+rendered+red%22+%7D%5D%3B%0A++++++++%7D%0A++++%7D++++%0A%7D%0A&style=native)

I would like to be able to uncomment `red_items.map(|item| {background_color: red, text: item.text})` somehow. If this is not possible, can I do it with Rust? What would be a minimally working example? What is the idiomatic way in Slint to do this? 

It seems like a common use case where you have a more low-level component and want to "sub-class" it to have more business-specific functionality.

I am coming from Svelte, where they have [reactive declarations](https://learn.svelte.dev/tutorial/reactive-declarations).

**Comment by ghost** [comment]
This is something that can be done in native could. E.g. in Rust there is the [MapModel](https://releases.slint.dev/1.7.2/docs/rust/slint/struct.mapmodel) than can be used to solve this problem.

**Reply by otiv-willem-vanhulle** [reply]
You mean "something that could be done in Rust"? How is this problem solved in practice for complex hierarchical user interfaces where the low level component is used in different ways by many different high level components?

**Reply by tronical** [reply]
In parallel, we're tracking https://github.com/slint-ui/slint/issues/1328 to add support for this on the Slint side. Suggestions for syntax are very welcome :)

**Reply by otiv-willem-vanhulle** [reply]
@tronical, I also noticed that Issue but there has not been any recent activity. How should I apply the mapmodel mentioned by @FloVanGH to the example I mentioned above?

**Reply by tronical** [reply]
What you could do is define a global helper like this (untested):

```slint
export global HighLevelToLowerLevelHelper {
    callback map-model([HighLevelStruct]) -> [LowlLevelStruct];
}
```

and then use it like this in Slint:

```slint
in property <[HighLevelStruct]> red_items;

out property<[LowLevelStruct]> generic_red_computed_items: HighLevelToLowerLevelHelper.map-model(red_items);
```

and finally in Rust associate the callback with the helper:

```rust
// Might have to use slint::MapModelExt;
ui.global::<HighLevelToLowerLevelHelper>().on_map_model(|hl_model| hl_model.map(|entry| LowLevelStruct { ... }));
```

If the mapping requires additional parameters, pass them along with the callback.
